src/Microsoft.AspNetCore.Razor.Language/Extensions/MetadataAttributePass.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Razor.Language.Extensions
+{
+    internal class MetadataAttributePass : IntermediateNodePassBase, 
+        
+        // Optimization pass is the best choice for this class. It's not an optimization, but it also doesn't add semantically
+        // meaningful information.
+        IRazorOptimizationPass
+    {
+        protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+        {
+            if (documentNode.Options == null || documentNode.Options.SuppressMetadataAttributes)
+            {
+                // Metadata attributes are turned off (or options not populated), nothing to do.
+                return;
+            }
+
+            // We need to be able to compute the data we need for the [RazorCompiledItem] attribute - that includes
+            // a full type name, and a document kind, and optionally an identifier.
+            // 
+            // If we can't use [RazorCompiledItem] then we don't care about the rest of the attributes.
+            var @namespace = documentNode.FindPrimaryNamespace();
+            if (@namespace == null || string.IsNullOrEmpty(@namespace.Content))
+            {
+                // No namespace node or it's incomplete. Skip.
+                return;
+            }
+
+            var @class = documentNode.FindPrimaryClass();
+            if (@class == null || string.IsNullOrEmpty(@class.ClassName))
+            {
+                // No class node or it's incomplete. Skip.
+                return;
+            }
+
+            if (documentNode.DocumentKind == null)
+            {
+                // No document kind. Skip.
+                return;
+            }
+
+            var identifier = codeDocument.GetIdentifier();
+            if (identifier == null)
+            {
+                // No identifier. Skip
+                return;
+            }
+
+            // [RazorCompiledItem] is an [assembly: ... ] attribute, so it needs to be applied at the global scope.
+            documentNode.Children.Insert(0, new RazorCompiledItemAttributeIntermediateNode()
+            {
+                TypeName = @namespace.Content + "." + @class.ClassName,
+                Kind = documentNode.DocumentKind,
+                Identifier = identifier,
+            });
+
+            // Now we need to add a [RazorSourceChecksum] for the source and for each import
+            // these are class attributes, so we need to find the insertion point to put them
+            // right before the class.
+            var insert = (int?)null;
+            for (var j = 0; j < @namespace.Children.Count; j++)
+            {
+                if (object.ReferenceEquals(@namespace.Children[j], @class))
```

Comment:
Should we just add an intermediate node collection called `Attributes` to our `ClassDeclarationIntermediateNode`? Could even add it to the `MemberDeclrationIntermediateNode` if we want to enable it everywhere 😉

Reply:
I started there but it's really a lot of work. Modeling attributes using a node has some challenges. You could do by adding the collection and not adding an Attribute node, but meh.